internal/users/localentries/localgroups.go

```diff
@@ -6,250 +6,291 @@ import (
-			return err
+
+	getGroupByName := func(name string) *types.GroupEntry {
+		idx := slices.IndexFunc(allGroups, func(g types.GroupEntry) bool { return g.Name == name })
+		if idx == -1 {
+			return nil
 		}
+		return &allGroups[idx]
 	}
 
-	return nil
-}
-
-// getPasswdUsernames gets the passwd entries and returns their usernames.
-func getPasswdUsernames() ([]string, error) {
-	var usernames []string
-	entries, err := GetPasswdEntries()
-	if err != nil {
-		return nil, err
+	for _, g := range groupsToRemove {
+		group := getGroupByName(g)
+		if group == nil {
+			continue
+		}
+		group.Users = slices.DeleteFunc(group.Users, func(u string) bool {
+			return u == username
+		})
 	}
-	for _, e := range entries {
-		usernames = append(usernames, e.Name)
+	for _, g := range groupsToAdd {
+		group := getGroupByName(g)
+		if group == nil {
+			continue
+		}
+		group.Users = append(group.Users, username)
 	}
 
-	return usernames, nil
+	return saveLocalGroups(ctx, allGroups)
 }
 
-// existingLocalGroups returns which groups from groupPath the user is part of.
-func existingLocalGroups(user, groupPath string) (groups []string, err error) {
+func parseLocalGroups(groupPath string) (groups []types.GroupEntry, invalidEntries []*string, err error) {
 	defer decorate.OnError(&err, "could not fetch existing local group")
 
-	localGroupsMu.RLock()
-	defer localGroupsMu.RUnlock()
+	log.Debugf(context.Background(), "Reading groups from %q", groupPath)
+
 	f, err := os.Open(groupPath)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	defer f.Close()
 
 	// Format of a line composing the group file is:
 	// group_name:password:group_id:user1,…,usern
 	scanner := bufio.NewScanner(f)
-	for scanner.Scan() {
+	for line := 0; scanner.Scan(); line++ {
 		t := strings.TrimSpace(scanner.Text())
-		if t == "" {
+		if t == "" || t[0] == '#' {
+			invalidEntries = append(invalidEntries, &t)
 			continue
 		}
-		elems := strings.Split(t, ":")
-		if len(elems) != 4 {
-			return nil, fmt.Errorf("malformed entry in group file (should have 4 separators): %q", t)
+
+		elems := strings.SplitN(t, ":", 4)
+		if len(elems) < 4 {
+			log.Errorf(context.Background(),
+				"Malformed entry in group file (should have 4 separators, got %d): %q", len(elems), t)
+			invalidEntries = append(invalidEntries, &t)
+			continue
 		}
 
-		n := elems[0]
-		users := strings.Split(elems[3], ",")
-		if !slices.Contains(users, user) {
+		name, passwd, gidValue, usersValue := elems[0], elems[1], elems[2], elems[3]
+
+		gid, err := strconv.ParseUint(gidValue, 10, 0)
+		if err != nil || gid > math.MaxUint32 {
+			log.Errorf(context.Background(),
+				"Failed parsing entry %q, unexpected GID value: %v", t, err)
+			invalidEntries = append(invalidEntries, &t)
 			continue
 		}
 
-		groups = append(groups, n)
+		var users []string
+		if usersValue != "" {
+			users = strings.Split(usersValue, ",")
+		}
+
+		invalidEntries = append(invalidEntries, nil)
```

Comment:
that's a bit surprising. I made a suggestion for a bit cleaner approach in https://github.com/3v1n0/authd/pull/9

Reply:
Yeah, I wanted to avoid the extra type, but I initially did the same in some lost commit